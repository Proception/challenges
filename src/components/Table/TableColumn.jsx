import React from 'react';

const Column = ({value}) => {
  if (Array.isArray(value)) {
    return (
      <div className={'table-column'}>{value.join(', ')}</div>
    );
  }

  if (!Array.isArray(value)) {
    return (
      <div className={'table-column'}>{value}</div>
    );
  }
};

export default Column;
