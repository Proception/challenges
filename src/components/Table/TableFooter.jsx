import React from 'react';

const TableFooter = ({ key, columns, handlePageNumbers, currentPage }) => {
  return (
    <div key={key} className={'table-footer'}>
      {
        columns
          .map(
            (columnValue) => 
              <div className={'table-footer-column'}>
                {
                  Array
                    .isArray(columnValue) ? 
                    columnValue.map(pageNumber => <a className={`page-navigators page-number-spacing ${currentPage === Number(pageNumber) ? 'active-page' : ''}`} value={pageNumber} onClick={handlePageNumbers}>{pageNumber}</a>) : 
                    columnValue
                }
              </div>
          )
      }
    </div>
  );
};

export default TableFooter;
