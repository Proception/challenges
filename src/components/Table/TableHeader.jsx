import React from 'react';
import PropTypes from 'prop-types';
import Column from './TableColumn.jsx';

const TableHeader = ({ columns }) => {
  return (
    <div className={'table-header'}>
      {
        columns.map((columnValue) => <Column value={columnValue} />)
      }
    </div>
  );
};


TableHeader.propTypes = {
  columns: PropTypes.array
};

export default TableHeader;
