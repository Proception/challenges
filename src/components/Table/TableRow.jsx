import React, {Fragment} from 'react';
import PropTypes from 'prop-types';
import Column from './TableColumn.jsx';

/**
 * @desc This is a reusable table row component.
 * @param {Array} rows : This is a prop passed to the table rows component for the row values
 * @returns {JSX}
 */
const TableRow = ({ rows }) => {
  if (rows.length === 0) {
    const noDataMessage = 'There is no consent data to display';
    return <div className={'no-data-message'}>{noDataMessage}</div>;
  } 
  
  return (
    <Fragment>
      {
        rows.map(row => {
          const rowValues = Object.values(row);
          return <div className={'table-row'}>
            {
              rowValues.map( rowvalue => <Column value={rowvalue} />)
            }
          </div>;
        })
      }
    </Fragment>
  );
};

TableRow.propTypes = {
  rows: PropTypes.array
};

export default TableRow;
