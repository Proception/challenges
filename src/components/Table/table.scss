.table-row {
  display: inline-flex;
  background-color: #F7F7F7;
}

.table-column {
  display: flex;
  padding: 10px;
  width: 250px;
  justify-content: center;
  text-align: center;
  font-size: 14px;
  color: #515050;
}

.table-header {
  display: inline-flex;
  font-weight: bolder;
  border-bottom: 1px solid #838486;
}

.table-container {
  display: block;
  text-align: center;
  margin: 20px;
}

.table-footer-column {
  display: flex;
  padding: 10px;
  width: 250px;
  justify-content: center;
  text-align: center;
}

.table-footer {
  display: inline-flex;
  color: #2867B2;
}

.no-data-message {
  display: inline-flex;
  padding: 20px;
}

.page-navigators {
  cursor: pointer;
  color: #2867B2;
  text-decoration: underline;
  font-size: 14px;
}

.page-navigators:hover {
  color: #515050;
}

.page-number-spacing {
  margin-right: 10px;
}

.active-page {
  cursor: not-allowed;
  color: #515050 !important; 
}