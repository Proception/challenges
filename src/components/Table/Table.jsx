import React, {Component} from 'react';
import PropTypes from 'prop-types';
import TableHeader from './TableHeader.jsx';
import TableRow from './TableRow.jsx';
import TableFooter from './TableFooter.jsx';

import './table.scss';
/**
 * @desc This is a reusable table component
 * @param {Array} tableRows : This is a prop passed to the table component with the table rows
 * @param {Array} tableHeaders : This contains the menu ID, action, and value
 * @returns {JSX}
 */
export default class Table extends Component {

  constructor () {
    super();
    this.state = {
      currentPage: 1,
      pages: [],
      displayRows: []
    };
  }

  componentDidMount() {
    this.formatPages();
  }

  handlePreviousPage = (e) => {
    const {currentPage} = this.state;
    this.displayPage(Number(currentPage) - 1);
  }

  handleNextPage = (e) => {
    const {currentPage} = this.state;
    this.displayPage(Number(currentPage) + 1);
  }

  handlePageNumbers = (e) => {
    const {currentPage} = this.state;
    const pageNumber = e.target.getAttribute('value');

    if (pageNumber !== currentPage) {
      this.displayPage(pageNumber);
    }
  }

  displayPage = (pageNumber) => {
    const {pages} = this.state;
    const displayPage = pages.filter(page => page[pageNumber]);
    this.setState({
      displayRows: displayPage[0][pageNumber],
      currentPage: Number(pageNumber)
    });
  }

  formatPages = () => {
    const {tableRows:consents} = this.props;
    const pageDataArray = [];
    const consentsCountLessThanThree = consents.length < 3;
    const consentsCountEqualToThree = consents.length === 3;
    const consentsCountGreaterThanThree = consents.length > 3;

    if (consentsCountLessThanThree) {
      pageDataArray[0] = {1: [consents[0], consents[1] ? consents[1] : '']};
    }

    if (consentsCountEqualToThree) {
      pageDataArray[0] = {1: [consents[0], consents[1] ? consents[1] : '']};
      pageDataArray[1] = {2: [consents[2]]};
    }

    if (consentsCountGreaterThanThree) {
      const pages = consents.length;
      let pageNumber = 1;
      for (let index = 0; index < pages;) {
        const key = pageNumber;
        const value = [consents[index], consents[index + 1] ? consents[index + 1] : ''];
        const pageData = {};
        pageData[key] = value;
        pageDataArray[pageNumber - 1] = pageData;
        index = index + 2;
        pageNumber++;
      }
    }

    if (consents.length > 0) {
      this.setState(prevState => {
        return {
          ...prevState,
          pages: pageDataArray
        };
      }, () => {
        if (this.state.displayRows.length === 0) {
          this.displayPage(1);
        }
      }
      );
    }
  }

  renderTableHeader = () => {
    const {tableHeaders} = this.props;
    return <TableHeader columns={tableHeaders}/>;
  }

  renderTableBody = () => {
    const {displayRows} = this.state;
    return <div className={'table-body'}><TableRow rows={displayRows}/></div>;
  }

  renderTableFooter = () => {
    const {currentPage, pages} = this.state;
    const remainingPages = pages.length - currentPage;
    const nextPage = <a className={'page-navigators'} onClick={this.handleNextPage}>
      { pages.length > 1 && remainingPages > 0 ? 'Next page  >>' : ''}
    </a>;
    const previousPage = <a className={'page-navigators'} onClick={this.handlePreviousPage}>
      { currentPage > 1 ? '<<  Previous page' : ''}
    </a>;
    const pageNumbers = pages.map(page => Object.keys(page));

    return <div className={'table-footer'}>
      <TableFooter currentPage={currentPage} handlePageNumbers={this.handlePageNumbers} columns={
        [
          previousPage, 
          pageNumbers, 
          nextPage
        ]}/>
    </div>;
  }

  render() {
    return <div className={'table-container'}>
      {
        this.renderTableHeader()
      }
      {
        this.renderTableBody()
      }
      {
        this.renderTableFooter()
      }
    </div>;
  }
}

Table.propTypes = {
  tableRows: PropTypes.array,
  tableHeaders: PropTypes.array
};
