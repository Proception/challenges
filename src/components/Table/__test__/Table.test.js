import React from 'react';
import sinon from 'sinon';
import renderer from 'react-test-renderer';
import TableColumn from '../TableColumn.jsx';
import TableRow from '../TableRow.jsx';
import TableHeader from '../TableHeader.jsx';
import TableFooter from '../TableFooter.jsx';
import Table from '../Table.jsx';
import {
  pageNumberEvent,
  preventDefault
} from '../../../../__mocks__/tableMocks';


describe('Table component', () => {
  describe('TableColumn component', () => {
    it('should render column correctly if a single value is passed', () => {
      const component = renderer.create(
        <TableColumn value={'beans'}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render column correctly if an array is passed', () => {
      const component = renderer.create(
        <TableColumn value={['beans', 'rice']}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render column with expected value', () => {
      const columnData = 'beans';
      const tree = renderer.create(
        <TableColumn value={columnData}/>,
      );

      const instance = tree.root;
      const columnValue = instance.findAllByProps({className: 'table-column'});
      expect(columnValue.length).toEqual(1);
      expect(columnValue[0].children[0]).toEqual(columnData);
    });
    it('should render column with expected array values', () => {
      const columnData = ['beans', 'rice'];
      const tree = renderer.create(
        <TableColumn value={columnData}/>,
      );

      const instance = tree.root;
      const columnValue = instance.findAllByProps({className: 'table-column'});
      expect(columnValue.length).toEqual(1);
      expect(columnValue[0].children).toEqual([`${columnData[0]}, ${columnData[1]}`]);
    });
  });
  describe('TableRow component', () => {
    it('should render no data message if no row data is passed', () => {
      const component = renderer.create(
        <TableRow rows={[]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render row correctly if row data is passed', () => {
      const rowData = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: [
          'Receive Newsletter', 'Be shown targeted ads'
        ]}
      ];
      const component = renderer.create(
        <TableRow rows={[...rowData]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('TableHeader component', () => {
    it('should render table headers', () => {
      const component = renderer.create(
        <TableHeader columns={['Name', 'Email', 'Consent given for']}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('TableFooter component', () => {
    it('should render table footer correctly', () => {
      const component = renderer.create(
        <TableFooter columns={['<< Previous page', [1,2,3], 'Next page >>']} handlePageNumbers={() => {}} currentPage= {1}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Table component', () => {
    it('should render table component with excatly 3 consents', () => {
      const tableHeaders = ['Name', 'Email', 'Consent given for'];
      const consents = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: ['Receive Newsletter','Be shown targeted ads']},
        {name: 'Omasan',email: 'benedict.esimaje@gmail.com',consents: ['Be shown targeted ads']},
        {name: 'Micheal',email: 'omasan@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']}
      ];
      const component = renderer.create(
        <Table tableHeaders={tableHeaders} tableRows={[...consents]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render table component with more than 3 consents', () => {
      const tableHeaders = ['Name', 'Email', 'Consent given for'];
      const consents = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: ['Receive Newsletter','Be shown targeted ads']},
        {name: 'Omasan',email: 'benedict.esimaje@gmail.com',consents: ['Be shown targeted ads']},
        {name: 'Micheal',email: 'omasan@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']},
        {name: 'Micheal',email: 'oman@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']}
      ];
      const component = renderer.create(
        <Table tableHeaders={tableHeaders} tableRows={[...consents]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render table component with less than 3 consents', () => {
      const tableHeaders = ['Name', 'Email', 'Consent given for'];
      const consents = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: ['Receive Newsletter','Be shown targeted ads']},
        {name: 'Omasan',email: 'benedict.esimaje@gmail.com',consents: ['Be shown targeted ads']}
      ];
      const component = renderer.create(
        <Table tableHeaders={tableHeaders} tableRows={[...consents]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should navigate to selected page using page numbers', () => {
      const tableHeaders = ['Name', 'Email', 'Consent given for'];
      const consents = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: ['Receive Newsletter','Be shown targeted ads']},
        {name: 'Omasan',email: 'benedict.esimaje@gmail.com',consents: ['Be shown targeted ads']},
        {name: 'Micheal',email: 'omasan@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']},
        {name: 'Micheal',email: 'oman@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']}
      ];
      const wrapper = shallow(<Table tableHeaders={tableHeaders} tableRows={[...consents]} />);
      sinon.spy(wrapper.instance(), 'handlePageNumbers');

      wrapper.instance().handlePageNumbers(pageNumberEvent);

      expect(wrapper.instance().handlePageNumbers.calledOnce)
        .toEqual(true);
      expect(wrapper.instance().handlePageNumbers.calledWith(pageNumberEvent)).toEqual(true);
    });
    it('should navigate to next page using handleNextPage and handlePreviousPage', () => {
      const tableHeaders = ['Name', 'Email', 'Consent given for'];
      const consents = [
        {name: 'Benedict', email: 'omasan.esimaje@gmail.com', consents: ['Receive Newsletter','Be shown targeted ads']},
        {name: 'Omasan',email: 'benedict.esimaje@gmail.com',consents: ['Be shown targeted ads']},
        {name: 'Micheal',email: 'omasan@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']},
        {name: 'Micheal',email: 'oman@gmail.com', consents: ['Receive Newsletter', 'Be shown targeted ads']}
      ];
      const wrapper = shallow(<Table tableHeaders={tableHeaders} tableRows={[...consents]} />);
      sinon.spy(wrapper.instance(), 'handleNextPage');
      sinon.spy(wrapper.instance(), 'handlePreviousPage');

      wrapper.instance().handleNextPage();
      wrapper.instance().handlePreviousPage();

      expect(wrapper.instance().handleNextPage.calledOnce)
        .toEqual(true);
      expect(wrapper.instance().handlePreviousPage.calledOnce)
        .toEqual(true);
    });
  });

});
