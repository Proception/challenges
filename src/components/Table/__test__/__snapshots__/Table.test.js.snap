// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component Table component should render table component with excatly 3 consents 1`] = `
<div
  className="table-container"
>
  <div
    className="table-header"
  >
    <div
      className="table-column"
    >
      Name
    </div>
    <div
      className="table-column"
    >
      Email
    </div>
    <div
      className="table-column"
    >
      Consent given for
    </div>
  </div>
  <div
    className="table-body"
  >
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Benedict
      </div>
      <div
        className="table-column"
      >
        omasan.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Receive Newsletter, Be shown targeted ads
      </div>
    </div>
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Omasan
      </div>
      <div
        className="table-column"
      >
        benedict.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Be shown targeted ads
      </div>
    </div>
  </div>
  <div
    className="table-footer"
  >
    <div
      className="table-footer"
    >
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators page-number-spacing active-page"
          onClick={[Function]}
          value={
            Array [
              "1",
            ]
          }
        >
          1
        </a>
        <a
          className="page-navigators page-number-spacing "
          onClick={[Function]}
          value={
            Array [
              "2",
            ]
          }
        >
          2
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          Next page  &gt;&gt;
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table component Table component should render table component with less than 3 consents 1`] = `
<div
  className="table-container"
>
  <div
    className="table-header"
  >
    <div
      className="table-column"
    >
      Name
    </div>
    <div
      className="table-column"
    >
      Email
    </div>
    <div
      className="table-column"
    >
      Consent given for
    </div>
  </div>
  <div
    className="table-body"
  >
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Benedict
      </div>
      <div
        className="table-column"
      >
        omasan.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Receive Newsletter, Be shown targeted ads
      </div>
    </div>
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Omasan
      </div>
      <div
        className="table-column"
      >
        benedict.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Be shown targeted ads
      </div>
    </div>
  </div>
  <div
    className="table-footer"
  >
    <div
      className="table-footer"
    >
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators page-number-spacing active-page"
          onClick={[Function]}
          value={
            Array [
              "1",
            ]
          }
        >
          1
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table component Table component should render table component with more than 3 consents 1`] = `
<div
  className="table-container"
>
  <div
    className="table-header"
  >
    <div
      className="table-column"
    >
      Name
    </div>
    <div
      className="table-column"
    >
      Email
    </div>
    <div
      className="table-column"
    >
      Consent given for
    </div>
  </div>
  <div
    className="table-body"
  >
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Benedict
      </div>
      <div
        className="table-column"
      >
        omasan.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Receive Newsletter, Be shown targeted ads
      </div>
    </div>
    <div
      className="table-row"
    >
      <div
        className="table-column"
      >
        Omasan
      </div>
      <div
        className="table-column"
      >
        benedict.esimaje@gmail.com
      </div>
      <div
        className="table-column"
      >
        Be shown targeted ads
      </div>
    </div>
  </div>
  <div
    className="table-footer"
  >
    <div
      className="table-footer"
    >
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators page-number-spacing active-page"
          onClick={[Function]}
          value={
            Array [
              "1",
            ]
          }
        >
          1
        </a>
        <a
          className="page-navigators page-number-spacing "
          onClick={[Function]}
          value={
            Array [
              "2",
            ]
          }
        >
          2
        </a>
      </div>
      <div
        className="table-footer-column"
      >
        <a
          className="page-navigators"
          onClick={[Function]}
        >
          Next page  &gt;&gt;
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table component TableColumn component should render column correctly if a single value is passed 1`] = `
<div
  className="table-column"
>
  beans
</div>
`;

exports[`Table component TableColumn component should render column correctly if an array is passed 1`] = `
<div
  className="table-column"
>
  beans, rice
</div>
`;

exports[`Table component TableFooter component should render table footer correctly 1`] = `
<div
  className="table-footer"
>
  <div
    className="table-footer-column"
  >
    &lt;&lt; Previous page
  </div>
  <div
    className="table-footer-column"
  >
    <a
      className="page-navigators page-number-spacing active-page"
      onClick={[Function]}
      value={1}
    >
      1
    </a>
    <a
      className="page-navigators page-number-spacing "
      onClick={[Function]}
      value={2}
    >
      2
    </a>
    <a
      className="page-navigators page-number-spacing "
      onClick={[Function]}
      value={3}
    >
      3
    </a>
  </div>
  <div
    className="table-footer-column"
  >
    Next page &gt;&gt;
  </div>
</div>
`;

exports[`Table component TableHeader component should render table headers 1`] = `
<div
  className="table-header"
>
  <div
    className="table-column"
  >
    Name
  </div>
  <div
    className="table-column"
  >
    Email
  </div>
  <div
    className="table-column"
  >
    Consent given for
  </div>
</div>
`;

exports[`Table component TableRow component should render no data message if no row data is passed 1`] = `
<div
  className="no-data-message"
>
  There is no consent data to display
</div>
`;

exports[`Table component TableRow component should render row correctly if row data is passed 1`] = `
<div
  className="table-row"
>
  <div
    className="table-column"
  >
    Benedict
  </div>
  <div
    className="table-column"
  >
    omasan.esimaje@gmail.com
  </div>
  <div
    className="table-column"
  >
    Receive Newsletter, Be shown targeted ads
  </div>
</div>
`;
